include/nbl/video/IGPUCommandBuffer.h

```diff
@@ -227,6 +232,93 @@ class IGPUCommandBuffer :
             for (uint32_t i = _first + 1u; i < IGPUPipelineLayout::DESCRIPTOR_SET_COUNT; ++i)
                 _destPplnLayouts[i] = nullptr;
     }
+
+    template <typename Cmd, typename... Args>
+    Cmd* emplace(Args&&... args)
+    {
+        if (m_segmentListTail == nullptr)
+        {
+            void* cmdSegmentMem = m_cmdpool->m_commandSegmentPool.allocate(IGPUCommandPool::COMMAND_SEGMENT_SIZE, alignof(IGPUCommandPool::CommandSegment));
+            if (!cmdSegmentMem)
+                return nullptr;
+
+            m_segmentListTail = new (cmdSegmentMem) IGPUCommandPool::CommandSegment;
+            m_segmentListHeadItr.m_segment = m_segmentListTail;
+        }
+
+        // Cmd* cmd = m_segmentListTail->allocate<Cmd, Args...>(args...);
+        Cmd* cmd = m_segmentListTail->allocate<Cmd>(std::forward<Args>(args)...);
+        if (!cmd)
+        {
+            void* nextSegmentMem = m_cmdpool->m_commandSegmentPool.allocate(IGPUCommandPool::COMMAND_SEGMENT_SIZE, alignof(IGPUCommandPool::CommandSegment));
+            if (nextSegmentMem == nullptr)
+                return nullptr;
+
+            IGPUCommandPool::CommandSegment* nextSegment = new (nextSegmentMem) IGPUCommandPool::CommandSegment;
+
+            // cmd = m_segmentListTail->allocate<Cmd, Args...>(args...);
+            cmd = m_segmentListTail->allocate<Cmd>(std::forward<Args>(args)...);
+            if (!cmd)
+                return nullptr;
+
+            m_segmentListTail->params.m_next = nextSegment;
+            m_segmentListTail = m_segmentListTail->params.m_next;
+        }
+
+        if (m_segmentListHeadItr.m_cmd == nullptr)
+            m_segmentListHeadItr.m_cmd = cmd;
+
+        return cmd;
+    }
+
+    IGPUCommandPool::CommandSegment::Iterator m_segmentListHeadItr = {};
+    IGPUCommandPool::CommandSegment* m_segmentListTail = nullptr;
+
+    private:
+        void deleteCommandSegmentList()
+        {
+            IGPUCommandPool::CommandSegment::Iterator itr = m_segmentListHeadItr;
+
+            if (itr.m_segment && itr.m_cmd)
+            {
+                bool lastCmd = itr.m_cmd->m_size == 0u;
+                while (!lastCmd)
+                {
+                    IGPUCommandPool::ICommand* currCmd = itr.m_cmd;
+                    IGPUCommandPool::CommandSegment* currSegment = itr.m_segment;
+
+                    itr.m_cmd = reinterpret_cast<IGPUCommandPool::ICommand*>(reinterpret_cast<uint8_t*>(itr.m_cmd) + currCmd->m_size);
+                    currCmd->~ICommand();
+                    // No need to deallocate currCmd because it has been allocated from the LinearAddressAllocator where deallocate is a No-OP and the memory will
+                    // get reclaimed in ~LinearAddressAllocator
+
+                    if ((reinterpret_cast<uint8_t*>(itr.m_cmd) - reinterpret_cast<uint8_t*>(itr.m_segment)) > IGPUCommandPool::CommandSegment::STORAGE_SIZE)
```

Comment:
Doing this 
```C++
for (auto cmdIt=segment->begin(); cmdIt!=segment->end(); cmdIt++)
        cmdIt->~ICommand();
```
is a bit tricky, because to implement the command-scrolling-pointer-arithmetic in `Iterator::operator++` you would need to query the size of the current command which cannot be done by a simple `m_cmd->getSize()` because we actually call the destructor (`cmdIt->~ICommand()`) __before__ we do `cmdIt++`.

So to do this, I think our other option could be to make the `Iterator` something like
```C++
struct Iterator
{
    ICommand* m_cmd;
    uint32_t m_size; // cache the size here before the destructor gets called
};
```
but then consider what would happen in the `Iterator::operator++`
```C++
Iterator& operator++()
{
    m_cmd = reinterpret_cast<ICommand*>(reinterpret_cast<uint8_t*>(m_cmd) + m_size); // update m_cmd with the cached size
    // but now we have no guarantee that the memory at m_cmd is a valid ICommand object --it would not be
    // for the last command since wipeNextCommandSize doesn't actually create a ICommand object but just
    // sets all of its memory to 0
    m_size = m_cmd->getSize(); // UB
}
```
This also ties back to this comment https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r948153002 where I suggested to not use `wipeNextCommandSize` this way and actually construct a `ICommand` object for the last command and just set its size to 0.

Reply:
fixed up your iterator impl:
https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r1040890916

also have a cool way to make `segment->end()==segment->getNext()->begin()`
https://github.com/Devsh-Graphics-Programming/Nabla/pull/345/files#r1041213574

Reply:
Resolved.